### PR TITLE
clamping initialized weights

### PR DESCRIPTION
**Pull request by sohamparikh**
# ✨ Description

Add max and min values for initialized weights for better stability during training. Inspired by 4.2.2 in the [OLMoE paper](https://arxiv.org/pdf/2409.02060).

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Change A
2. Change B

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [ ] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [ ] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

**Review by tscholak — COMMENTED** [review]
LGTM, @jlamypoirier ?

File: fast_llm/tensor.py [hunk 1]
```diff
@@ -312,15 +312,15 @@ def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator)
 init_ones_ = init_fill_(1.0)
 
 
-def init_normal_(mean=0.0, std=1.0):
+def init_normal_(mean=0.0, std=1.0, min_val=None, max_val=None):
     def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
-        return tensor.normal_(mean, std, generator=generator)
+        return tensor.clamp(tensor.normal_(mean, std, generator=generator), min=min_val, max=max_val)
```
Comment (jlamypoirier) [inline fast_llm/tensor.py:317]: I don't think that works. Probably `tensor.normal_.clamp_`?
Also let's put in a `if` so there is no clamp call by default.

File: fast_llm/tensor.py [hunk 2]
```diff
@@ -312,15 +312,15 @@ def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator)
...
 
     return init_
 
 
-def init_uniform_(low=0.0, high=1.0):
+def init_uniform_(low=0.0, high=1.0, min_val=None, max_val=None):
     def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
-        return tensor.uniform_(low, high, generator=generator)  # noqa
+        return tensor.clamp(tensor.uniform_(low, high, generator=generator), min=min_val, max=max_val)  # noqa
```
Comment (jlamypoirier) [inline fast_llm/tensor.py:324]: Same here

File: fast_llm/layers/language_model/config.py [hunk 3]
```diff
@@ -167,6 +177,15 @@ class LanguageModelBaseConfig(LanguageModelArchitectureConfig, BaseModelConfig):
     def _validate(self):
         if self.transformer.init_method_std is None:
             self.transformer.init_method_std = self.transformer.hidden_size**-0.5
+        if self.transformer.init_method_max is None:
+            self.transformer.init_method_max = float("inf")
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:181]: Not needed

File: fast_llm/layers/language_model/config.py [hunk 4]
```diff
@@ -118,6 +118,16 @@ class LanguageModelBaseConfig(LanguageModelArchitectureConfig, BaseModelConfig):
         hint=FieldHint.feature,
         valid=check_field(Assert.geq, 0),
     )
+    init_method_max_embed: float = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:121]: `float|None`

**Review by jlamypoirier — COMMENTED** [review]
Looks mostly ok, some minor issues.

I'm also worried about the bloat. Ideally we'd want a common sub-config class for all initializations, but no need to address right now.

File: fast_llm/layers/language_model/config.py (hunk 4 above)
Comment (sohamparikh) [reply fast_llm/layers/language_model/config.py:121]: do you also want this change in `fast_llm/layers/transformer/config.py`?

File: fast_llm/layers/language_model/config.py (hunk 3 above)
Comment (sohamparikh) [reply fast_llm/layers/language_model/config.py:181]: same question as above

File: fast_llm/layers/language_model/config.py (hunk 4 above)
Comment (jlamypoirier) [reply fast_llm/layers/language_model/config.py]: Yes, all of them

File: fast_llm/tensor.py [hunk 5]
```diff
@@ -312,15 +312,21 @@ def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator)
 init_ones_ = init_fill_(1.0)
 
 
-def init_normal_(mean=0.0, std=1.0):
+def init_normal_(mean=0.0, std=1.0, min_val=None, max_val=None):
     def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
-        return tensor.normal_(mean, std, generator=generator)
+        if min_val is not None or max_val is not None:
+            return tensor.normal_(mean, std, generator=generator).clamp(min=min_val, max=max_val)
```
Comment (jlamypoirier) [inline fast_llm/tensor.py:318]: the `tensor.normal_` call doesn't need to be inside the loop

File: fast_llm/tensor.py [hunk 6]
```diff
@@ -312,15 +312,23 @@ def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator)
 init_ones_ = init_fill_(1.0)
 
 
-def init_normal_(mean=0.0, std=1.0):
+def init_normal_(mean=0.0, std=1.0, min_val=None, max_val=None):
     def init_(meta: ParameterMeta, tensor: torch.Tensor, generator: torch.Generator):  # noqa
-        return tensor.normal_(mean, std, generator=generator)
+        initial_tensor = tensor.normal_(mean, std, generator=generator)
+        if min_val is not None or max_val is not None:
+            return initial_tensor.clamp(min=min_val, max=max_val)  # noqa
```
Comment (jlamypoirier) [inline fast_llm/tensor.py:319]: This is an in-place method, so you need `clamp_`

File: fast_llm/layers/language_model/config.py [hunk 7]
```diff
@@ -118,6 +118,16 @@ class LanguageModelBaseConfig(LanguageModelArchitectureConfig, BaseModelConfig):
         hint=FieldHint.feature,
         valid=check_field(Assert.geq, 0),
     )
+    init_method_max_embed: float|None = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/config.py:121]: The spacing is wrong, it should be `float | None`. Did you set up pre-commit?

Comment (sohamparikh) [reply fast_llm/layers/language_model/config.py]: ah, thanks! Just setup pre-commit now

**Review by jlamypoirier — APPROVED** [review]

**Pull request merged**